servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/MulticastSingle.java

```diff
@@ -0,0 +1,371 @@
+                    terminalSubscriber.safeTerminateFromSource(subscriber);
+                    break;
+                } else {
+                    @SuppressWarnings("unchecked")
+                    Subscriber<? super T>[] newSubs = (Subscriber<? super T>[])
+                            Array.newInstance(Subscriber.class, currSubs.length + 1);
+                    System.arraycopy(currSubs, 0, newSubs, 0, currSubs.length);
+                    newSubs[currSubs.length] = multiSubscriber;
+                    if (newSubscribersUpdater.compareAndSet(this, currSubs, newSubs)) {
+                        // Note we invoke onSubscribe AFTER the subscribers array because the subscription methods
+                        // depend upon this state (cancellation needs the subscriber to be visible in the array). This
+                        // may result in onSuccess(t) or onError(t) being called before onSubscribe() which is
+                        // unexpected for the Subscriber API, but MulticastSubscriber can tolerate this and internally
+                        // queues signals.
+                        multiSubscriber.onSubscribe(() -> removeSubscriber(multiSubscriber));
+                        if (sCount == minSubscribers) {
+                            // This operator has special behavior where it chooses to use the AsyncContext and signal
+                            // offloader from the last subscribe operation.
+                            original.delegateSubscribe(this, contextMap, contextProvider);
+                        }
+                        break;
+                    }
+                }
+            }
+        }
+
+        void removeSubscriber(Subscriber<T> subscriber) {
+            for (;;) {
+                final Subscriber<? super T>[] currSubs = subscribers;
+                final int i;
+                if (currSubs.length == 0 || currSubs.length == 1 && currSubs[0] instanceof TerminalSubscriber ||
+                        (i = ArrayUtils.indexOf(subscriber, currSubs)) < 0) {
+                    return;
+                }
+                @SuppressWarnings("unchecked")
+                Subscriber<? super T>[] newSubs = (Subscriber<? super T>[])
+                        Array.newInstance(Subscriber.class, currSubs.length - 1);
+                if (i == 0) {
+                    System.arraycopy(currSubs, 1, newSubs, 0, newSubs.length);
+                } else {
+                    System.arraycopy(currSubs, 0, newSubs, 0, i);
+                    System.arraycopy(currSubs, i + 1, newSubs, i, newSubs.length - i);
+                }
+                if (newSubscribersUpdater.compareAndSet(this, currSubs, newSubs)) {
+                    if (cancelUpstream && newSubs.length == 0) {
+                        // Reset the state when all subscribers have cancelled to allow for re-subscribe.
+                        state = new State();
+                        delayedCancellable.cancel();
```

Comment:
Consider adding a test when a new subscriber re-subscribes after all previous subscribers canceled

Reply:
This should be covered in `singleSubscriberCancelStillDeliversData`